### PR TITLE
fix: check origin of user

### LINKED ISSUE #2576: SCIM user creation works when providing invalid origin
### What version of UAA are you running?

76.21.0 , issue is also present in older versions

### How are you deploying the UAA?

Via bosh release. However, issue is independent of deployment

### What did you do?

Create or update a shadow user with invalid origin.

```
curl -X POST 'https://<uaaurl>/Users' -H "Authorization: bearer <admintoken>" -H "Content-type: application/json" -d '{"userName": "user-does-not-exist",  "origin": "does-not-exist", "emails": [{"value": [test@email.com](mailto:test@email.com), "primary": false}]}'
```

### What did you expect to see? What goal are you trying to achieve with the UAA?

Creation or update is rejected because the origin does not map to a known identity provider.

### What did you see instead?

User is created or updated.

### PR DESCRIPTION
**Pull request by klaus-sap**
issue: https://github.com/cloudfoundry/uaa/issues/2576
check must be switched on in idz configuration to avoid incompatible change

### CODE REVIEW

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java [hunk 1]
```diff
@@ -1048,26 +1053,66 @@ void updateLastLogonTime() {
...
-        for (int i = 2; i < 12; i++) {
+        try {
+            idzManager.getCurrentIdentityZone().getConfig().getUserConfig().setMaxUsers(10);
+            for (int i = 1; i < 12; i++) {
+                scimUser.setId("user-id-" + i);
+                scimUser.setUserName("user" +i+ "@example.com");
+                scimUser.setPassword(randomString());
+                scimUser = jdbcScimUserProvisioning.create(scimUser, IdentityZoneHolder.getCurrentZoneId());
+            }
+        } catch (InvalidScimResourceException e) {
```
Comment (adrianhoelzl-sap) [inline server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java]: Either use `assertThrows` or add `fail` before this line

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java [hunk 2]
```diff
@@ -1048,26 +1053,66 @@ void updateLastLogonTime() {
...
+        ScimUser scimUser = new ScimUser("user-id-"+randomId, "user@example.com", "User", "Example");
+        ScimUser.Email email = new ScimUser.Email();
+        email.setValue("user-"+randomId+"@example.com");
+        scimUser.setEmails(Collections.singletonList(email));
+        scimUser.setPassword(randomString());
+        scimUser.setOrigin(invalidOrigin);
+        try {
+            idzManager.getCurrentIdentityZone().getConfig().getUserConfig().setCheckOriginEnabled(true);
+            scimUser = jdbcScimUserProvisioning.create(scimUser, IdentityZoneHolder.getCurrentZoneId());
+        } catch (InvalidScimResourceException e) {
```
Comment (adrianhoelzl-sap) [inline server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java]: Same here

Comment (klaus-sap) [reply server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java]: fixed

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java (hunk 1 above)
Comment (klaus-sap) [reply server/src/test/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioningTests.java]: fixed

**Comment by strehle** [comment]
see sonar issues, https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2688

**Review by strehle — COMMENTED** [review]
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2688

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 3]
```diff
@@ -271,6 +274,9 @@ public ScimUser update(final String id, final ScimUser user, String zoneId) thro
         logger.debug("Updating user " + user.getUserName());
         final String origin = hasText(user.getOrigin()) ? user.getOrigin() : OriginKeys.UAA;
         user.setOrigin(origin);
+        if (isCheckOriginEnabled(zoneId)) {
+            checkOrigin(origin, zoneId);
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java:281]: sonar does not show coverage for this here, can you please add a test and I would prefere an IT for negative testing as well

Comment (klaus-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java:281]: There is already a negative test for create: cannotCreateUserWithInvalidOrigin

Comment (klaus-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java:281]: Added test cannotUpdateUserWithInvalidOrigin

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 4]
```diff
@@ -588,7 +588,8 @@ private boolean isCheckOriginEnabled(String zoneId) {
 
     private void checkOrigin(String origin, String zoneId) {
         Integer count = jdbcTemplate.queryForObject("select count(*) from identity_provider where origin_key=? and identity_zone_id=? ", Integer.class, origin, zoneId);
-        if (count == 0) {
+        int c = (count == null) ? 0 : count;
```
Comment (hsinn0) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: It seems that the query string is existing code. This query operation can be optimized by using `exists` instead of `count` though, because we just need to know whether there is at least 1 such record that meets the condition or not. For example, in mysql, the query can be as follows instead:

`select exists (select * from identity_provider where origin_key=? and identity_zone_id=?)`

And then the result can be assigned to a `boolean` variable, removing the need to check for null Integer value.

**Review by hsinn0 — DISMISSED** [review]

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java (hunk 4 above)
Comment (klaus-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: In this case, count can only be 0 or 1, and there is a matching index on table identity_provider.
https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/resources/org/cloudfoundry/identity/uaa/db/postgresql/V2_0_0__Multitenancy.sql#L37
So I don't think that the query can be optimized with respect to runtime.

Comment (hsinn0) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: @klaus-sap , do you not agree that exists query is more efficient than count query? Even if the count can be only 1 or 0, count query needs to go through all records while exists query does not, right?

Comment (hsinn0) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: I see. There is unique index on `identity_provider (identity_zone_id,origin_key)`. So in theory, not much to gain on the SQL side then. In that case, the only benefit of changing the query is likely not having to covert from Integer to int, and I am OK if you don't think that is important enough to make further change.

**Review by hsinn0 — APPROVED** [review]

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 5]
```diff
@@ -196,6 +196,9 @@ public ScimUser create(final ScimUser user, String zoneId) {
         if (!hasText(user.getOrigin())) {
             user.setOrigin(OriginKeys.UAA);
         }
+        if (isCheckOriginEnabled(zoneId)) {
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: This here is in createUser, where we have some lines before this call validateUserLimit , 
Both potentially perform a SQL to resolve the identityzone. Can we combine it and do the  identityZoneManager.getCurrentIdentityZone check once?

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java [hunk 6]
```diff
@@ -565,4 +571,26 @@ private void validateUserLimit(String zoneId) {
...
+            return userConfig.isCheckOriginEnabled();
+        } catch (ZoneDoesNotExistsException e) {
+            logger.debug("could not retrieve identity zone with id: {}", zoneId);
+        }
+        return false;
+    }
+
+    private void checkOrigin(String origin, String zoneId) {
+        Integer count = jdbcTemplate.queryForObject("select count(*) from identity_provider where origin_key=? and identity_zone_id=? ", Integer.class, origin, zoneId);
+        int c = (count == null) ? 0 : count;
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: line 591 + 592 -> if (count == null || count == 0) { throw xxx

Comment (klaus-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: fixed with commit https://github.com/cloudfoundry/uaa/pull/2688/commits/e382e49e6a53b04441332d272412036aee4e9260

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java (hunk 5 above)
Comment (klaus-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java]: fixed with [d9f8f25](https://github.com/cloudfoundry/uaa/pull/2688/commits/d9f8f25bfb0d791fcb5dfd8ea1d5c1e432c72428)

**Review by strehle — APPROVED** [review]
